src/python/pants/testutil/rule_runner.py

```diff
@@ -0,0 +1,238 @@
+from pants.util.dirutil import recursive_dirname, safe_file_dump, safe_mkdir, safe_open
+from pants.util.meta import frozen_after_init
+from pants.util.ordered_set import FrozenOrderedSet
+
+_P = TypeVar("_P")
+
+
+@dataclass(frozen=True)
+class GoalRuleResult:
+    exit_code: int
+    stdout: str
+    stderr: str
+
+    @staticmethod
+    def noop() -> "GoalRuleResult":
+        return GoalRuleResult(0, stdout="", stderr="")
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class RuleRunner:
+    build_root: str
+    build_config: BuildConfiguration
+    scheduler: SchedulerSession
+
+    def __init__(
+        self,
+        *,
+        rules: Optional[Iterable] = None,
+        target_types: Optional[Iterable[Type[Target]]] = None,
+        objects: Optional[Dict[str, Any]] = None,
+        context_aware_object_factories: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.build_root = os.path.realpath(mkdtemp(suffix="_BUILD_ROOT"))
+        safe_mkdir(self.build_root, clean=True)
+        safe_mkdir(self.pants_workdir)
+        BuildRoot().path = self.build_root
+
+        # TODO: Redesign rule registration for tests to be more ergonomic and to make this less
+        #  special-cased.
+        all_rules = (
+            *(rules or ()),
+            *source_root.rules(),
+            QueryRule(WrappedTarget, (Address, OptionsBootstrapper)),
+        )
+        build_config_builder = BuildConfiguration.Builder()
+        build_config_builder.register_aliases(
+            BuildFileAliases(
+                objects=objects, context_aware_object_factories=context_aware_object_factories
+            )
+        )
+        build_config_builder.register_rules(all_rules)
+        build_config_builder.register_target_types(target_types or ())
+        self.build_config = build_config_builder.create()
+
+        options_bootstrapper = OptionsBootstrapper.create(
+            env={}, args=["--pants-config-files=[]"], allow_pantsrc=False
+        )
+        global_options = options_bootstrapper.bootstrap_options.for_global_scope()
+        local_store_dir = global_options.local_store_dir
+        local_execution_root_dir = global_options.local_execution_root_dir
+        named_caches_dir = global_options.named_caches_dir
+
+        graph_session = EngineInitializer.setup_graph_extended(
+            pants_ignore_patterns=[],
+            use_gitignore=False,
+            local_store_dir=local_store_dir,
+            local_execution_root_dir=local_execution_root_dir,
+            named_caches_dir=named_caches_dir,
+            native=Native(),
+            options_bootstrapper=options_bootstrapper,
+            build_root=self.build_root,
+            build_configuration=self.build_config,
+            execution_options=ExecutionOptions.from_bootstrap_options(global_options),
+        ).new_session(build_id="buildid_for_test", should_report_workunits=True)
+        self.scheduler = graph_session.scheduler_session
+
```

Comment:
Note that we have no way of registering these clean up methods:

https://github.com/pantsbuild/pants/blob/2fc8a42dc1325b8851eedba4202d52f63e013613/src/python/pants/testutil/test_base.py#L243-L258

I don't think we need them, though. The tmpdir for `build_root` should be automatically cleaned. `BuildRoot().path` will get reset the next time that a new `RuleRunner` is created, and it can't impact outside of the test file thanks to v2's isolation.

Reply:
That sounds right to me.